### PR TITLE
`super` can only be used inside a short-hand method definition.

### PR DESCRIPTION
**Pull request by 0xe**

### CODE REVIEW

File: rhino/src/main/resources/org/mozilla/javascript/resources/Messages_zh_CN.properties [hunk 1]
```diff
@@ -432,6 +432,9 @@ msg.no.paren =\
 msg.reserved.id =\
     标识符是保留字: {0}
 
+msg.super.shorthand.function =\
```
Comment (0xe) [inline rhino/src/main/resources/org/mozilla/javascript/resources/Messages_zh_CN.properties:435]: image

I'm not sure why, but this should be displayed like this. Probably my fonts don't have the correct character sets? :shrug:

File: rhino/src/main/resources/org/mozilla/javascript/resources/Messages_fr.properties [hunk 2]
```diff
@@ -405,6 +405,9 @@ msg.no.paren =\
 msg.reserved.id =\
     l''identifiant est un mot r\u00E9serv\u00E9
 
+msg.super.shorthand.function =\
```
Comment (andreabergia) [inline rhino/src/main/resources/org/mozilla/javascript/resources/Messages_fr.properties:408]: Do you know French? Because otherwise, I would just ignore this translation

File: rhino/src/main/resources/org/mozilla/javascript/resources/Messages_zh_CN.properties (hunk 1 above)
Comment (andreabergia) [inline rhino/src/main/resources/org/mozilla/javascript/resources/Messages_zh_CN.properties:435]: Same as for french

File: rhino/src/main/java/org/mozilla/javascript/ast/FunctionNode.java [hunk 3]
```diff
@@ -82,6 +82,11 @@ public static enum Form {
     private int rp = -1;
     private boolean hasRestParameter;
 
+    @Override
```
Comment (andreabergia) [inline rhino/src/main/java/org/mozilla/javascript/ast/FunctionNode.java:85]: Why do you override this?

File: rhino/src/main/java/org/mozilla/javascript/ast/ScriptNode.java [hunk 4]
```diff
@@ -39,6 +39,16 @@ public class ScriptNode extends Scope {
...
 
+    public boolean isMethodDefinition() {
+        return false;
+    }
+
+    public void setMethodDefinition(boolean methodDefinition) {
+        isMethodDefinition = methodDefinition;
+    }
+
+    protected boolean isMethodDefinition;
```
Comment (andreabergia) [inline rhino/src/main/java/org/mozilla/javascript/ast/ScriptNode.java:50]: Can we make it `private` if we remove the override of the getter in `FunctionNode`?

File: rhino/src/main/java/org/mozilla/javascript/RhinoException.java [hunk 5]
```diff
@@ -135,13 +135,13 @@ final void recordErrorOrigin(
         if (sourceName != null) {
             initSourceName(sourceName);
         }
-        if (lineNumber != 0) {
+        if (lineNumber > 0) {
```
Comment (andreabergia) [inline rhino/src/main/java/org/mozilla/javascript/RhinoException.java:138]: Why these changes?

File: rhino/src/main/java/org/mozilla/javascript/IRFactory.java [hunk 6]
```diff
@@ -845,6 +845,11 @@ private Node transformLetNode(LetNode node) {
     }
 
     private Node transformLiteral(AstNode node) {
+        // See 15.4.2 Static Semantics: HasDirectSuper
+        if (node.getParent() instanceof FunctionCall
+                && node.getType() == Token.SUPER
+                && parser.currentScriptOrFn.isMethodDefinition())
+            parser.reportError("msg.super.shorthand.function");
```
Comment (andreabergia) [inline rhino/src/main/java/org/mozilla/javascript/IRFactory.java:852]: I'm not clear what this is supposed to do
